File: Manual/LightField/Add-in and Automation SDK/Samples/VisualBasic Add-Ins/File Sample/FileSample.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Markup
Imports System.AddIn
Imports System.IO
Imports System.AddIn.Pipeline
Imports PrincetonInstruments.LightField.AddIns

'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
'''''''''' Generate data in memory and stores the data in two SPE files at user-designed location. ''''''''''
'''''''''' One file contrains multiple frames with a single region,                                ''''''''''
'''''''''' and the other contains a single frame with multiple regions.                            ''''''''''
'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
Public Class FileSample
    Inherits AddInBase

    Public toolBarState_ As Boolean?

    Public ReadOnly Property UISupport As UISupport
        Get
            Return UISupport.DataToolBar
        End Get
    End Property

    Public Sub Activate(ByVal app As ILightFieldApplication)
        'Capture interface
        LightFieldApplication = app
        toolBarState_ = Nothing 'Forces it to be a button. True/False would make it a checkbox.
    End Sub

    Public Sub Deactivate()
    End Sub

    Public Overrides Sub UIDataToolBarExecute()
        GenerateFiles()
    End Sub

    Public Overrides ReadOnly Property UIDataToolBarTitle As String
        Get
            Return "File Sample"
        End Get
    End Property

    Public Overrides ReadOnly Property UIDataToolBarToolTip As String
        Get
            Return "Generate a multi-frame single region file and a multi region single frame file sample"
        End Get
    End Property

    Public Overrides ReadOnly Property UIDataToolBarIsEnabled As Boolean
        Get
            Return True
        End Get
    End Property

    Public Overrides Property UIDataToolBarIsChecked As Nullable(Of Boolean)
        Get
            Return toolBarState_
        End Get
        Set(value As Nullable(Of Boolean)) 'The button In LightField was pushed
            GenerateFiles()
        End Set
    End Property

    'The following routine demonstrates how to create files using the LightField Addin SDK. 
    'Create one File with 2 frames and 1 region  (TwoFrameOneRoi.Spe)
    'Create one File with 1 frames and 2 regions (OneFrameTwoRoi.Spe)        
    Private Sub GenerateFiles()
        Dim filemgr As IFileManager = LightFieldApplication.FileManager 'Get the addin file manager
        If filemgr Is Nothing Then
            Exit Sub
        End If

        'Part 1: two frames one region
        Dim w1() As Integer = {100}
        Dim h1() As Integer = {200}
        Dim frame1(100 * 200) As UShort
        Dim frame2(100 * 200) As UShort
        For pix As Integer = 0 To 100 * 200 - 1
            frame1(pix) = pix
            frame2(pix) = pix * 2
        Next
        'Folder Selector
        Dim dialog As Object = New System.Windows.Forms.FolderBrowserDialog()
        dialog.Description = "Please choose folder for resulting SPE file(s)."
        Dim result As System.Windows.Forms.DialogResult = dialog.ShowDialog()
        If result <> System.Windows.Forms.DialogResult.OK Then 'If do not want to write the files.
            Exit Sub
        End If
        Dim path As String = dialog.SelectedPath 'Local variable containing path
        'Create a file with 2 frames and the proper width and height.
        'The file is initially created as empty and the user must load data to it. 
        Dim roi As RegionOfInterest = New RegionOfInterest(0, 0, w1(0), h1(0), 1, 1)
        Dim rois As RegionOfInterest() = {roi}
        Dim TwoFrameOneRoi As IImageDataSet = filemgr.CreateFile(path + "\\TwoFrameOneRoi.spe", rois, 2, ImageDataFormat.MonochromeUnsigned16)
        'Put data to frame 1
        Dim data1 As IImageData = TwoFrameOneRoi.GetFrame(0, 0)
        data1.SetData(frame1)
        'Put data to frame 2
        Dim data2 As IImageData = TwoFrameOneRoi.GetFrame(0, 1)
        data2.SetData(frame2)
        'Finally close this file
        filemgr.CloseFile(TwoFrameOneRoi)

        'Part 2: one frame two regions
        Dim w2 As Integer() = {300, 500}
        Dim h2 As Integer() = {400, 600}
        Dim frame1_roi1(300 * 400) As UShort
        For pix As Integer = 0 To 300 * 400 - 1
            frame1_roi1(pix) = pix Mod 65536
        Next
        Dim frame1_roi2(500 * 600) As UShort
        For pix As Integer = 0 To 500 * 600 - 1
            frame1_roi2(pix) = pix Mod 65536
        Next
        'Create a file, get the buffer and fill it in for each region.
        'The file is initially created as empty and the user must load data to it. 
        'The regions can not overlap
        Dim r1 As RegionOfInterest = New RegionOfInterest(0, 0, w2(0), h2(0), 1, 1)
        Dim r2 As RegionOfInterest = New RegionOfInterest(w2(0) + 1, h2(0) + 1, w2(1), h2(1), 1, 1)
        Dim rois2 As RegionOfInterest() = {r1, r2}
        Dim OneFrameTwoRoi As IImageDataSet = filemgr.CreateFile(path + "\\OneFrameTwoRoi.spe", rois2, 1, ImageDataFormat.MonochromeUnsigned16)
        'Put data to region 1
        Dim data2_roi1 As IImageData = OneFrameTwoRoi.GetFrame(0, 0)
        data2_roi1.SetData(frame1_roi1)
        'Put data to region 2
        Dim data2_roi2 As IImageData = OneFrameTwoRoi.GetFrame(1, 0)
        data2_roi2.SetData(frame1_roi2)
        'Finally close this file
        filemgr.CloseFile(OneFrameTwoRoi)
    End Sub

End Class

File: Manual/LightField/Add-in and Automation SDK/Samples/CSharp Add-Ins/File Sample/FileSample.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.AddIn;
using System.AddIn.Pipeline;

using PrincetonInstruments.LightField.AddIns;

namespace LightFieldAddInSamples
{   
    ///////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////
    [AddIn("File Sample (C#)", Version = "1.0.0", Publisher = "Teledyne Princeton Instruments", Description = "Shows how to generate SPE files from raw data")]
    [QualificationData("IsSample", "True")]
    public class FileSample : AddInBase, ILightFieldAddIn
    {
        bool? toolBarState_;
        ///////////////////////////////////////////////////////////////////////
        public UISupport UISupport { get { return UISupport.DataToolBar; } }
        ///////////////////////////////////////////////////////////////////////
        public void Activate(ILightFieldApplication app)
        {
            // Capture Interface
            LightFieldApplication = app;
            toolBarState_ = null; // Forces it to be a button, true or false
                                  // would make it a checkbox
        }
        ///////////////////////////////////////////////////////////////////////
        public void Deactivate() { }
        ///////////////////////////////////////////////////////////////////////
        public override string UIDataToolBarTitle { get { return "File Sample"; } }
        ///////////////////////////////////////////////////////////////////////
        public override bool? UIDataToolBarIsChecked
        {
            get { return toolBarState_; }

            // The Button In LightField Was Pushed
            set { GenerateFiles(); }
        }
        ///////////////////////////////////////////////////////////////////////        
        public override string UIDataToolBarToolTip
        {
            get { return "Generate a multi-frame single region file and a multi region single frame file sample"; }
        }
        ///////////////////////////////////////////////////////////////////////        
        public override bool UIDataToolBarIsEnabled{ get { return true; } }
        ///////////////////////////////////////////////////////////////////////        
        public override void UIDataToolBarExecute()
        {
            GenerateFiles();
        }        
        ////////////////////////////////////////////////////////////////////////
        //  The following routine demonstrates how to create files using the 
        //  LightField Addin SDK. 
        //
        //  Create 1 File with 2 Frames and 1 Region  (TwoFrameOneRoi.Spe)
        //  Create 1 File with 1 Frames and 2 Regions (OneFrameTwoRoi.Spe)        
        ////////////////////////////////////////////////////////////////////////
        private void GenerateFiles()
        {
            // Get the addin file manager
            var filemgr = LightFieldApplication.FileManager;
            if (filemgr != null)
            {
                ///////////////////////////////////////////////////////////////
                // Part 1: Two Frames 1 Region
                ///////////////////////////////////////////////////////////////                
                int[] w1 = { 100 };
                int[] h1 = { 200 };

                // Pseudo Frame 1 Region 1, Frame 2 Region 1
                ushort[] frame1 = new ushort[100 * 200];
                ushort[] frame2 = new ushort[100 * 200];
                for (int pix = 0; pix < 100 * 200; pix++)
                {
                    frame1[pix] = (ushort)pix;
                    frame2[pix] = (ushort)(pix * 2);
                }

                // Folder Selector
                var dialog = new System.Windows.Forms.FolderBrowserDialog();
                dialog.Description = "Please choose folder for resulting SPE file(s).";
                System.Windows.Forms.DialogResult result = dialog.ShowDialog();

                // Something bad happened we don't want to write the files.
                if (result != System.Windows.Forms.DialogResult.OK)
                    return;

                // Local variable containing path 
                string path = dialog.SelectedPath;

                // Create a file with 2 frames and the proper width and height
                // The file is initially created as empty and the user must load data to it. 
                RegionOfInterest roi = new RegionOfInterest(0, 0, w1[0], h1[0], 1, 1);
                RegionOfInterest[] rois = { roi };
                IImageDataSet TwoFrameOneRoi = filemgr.CreateFile(path + "\\TwoFrameOneRoi.spe",
                                                          rois,
                                                          2,                  // Frames
                                                          ImageDataFormat.MonochromeUnsigned16);                

                // Put Data to frame 1
                IImageData data1 = TwoFrameOneRoi.GetFrame(0, 0);
                data1.SetData(frame1);

                // Put Data to frame 2
                IImageData data2 = TwoFrameOneRoi.GetFrame(0, 1);
                data2.SetData(frame2);
                
                // Finally close this file
                filemgr.CloseFile(TwoFrameOneRoi);

                ///////////////////////////////////////////////////////////////
                // Part 2: 1 Frame 2 Regions
                ///////////////////////////////////////////////////////////////                
                int[] w2 = { 300, 500 };
                int[] h2 = { 400, 600 };

                // Pseudo Frame 2 Region 1
                ushort[] frame1_roi1 = new ushort[300 * 400];
                for (int pix = 0; pix < 300 * 400; pix++)
                    frame1_roi1[pix] = (ushort)pix;

                // Pseudo Frame 2 Region 2
                ushort[] frame1_roi2 = new ushort[500 * 600];
                for (int pix = 0; pix < 500 * 600; pix++)
                    frame1_roi2[pix] = (ushort)pix;

                // Create a file, get the buffer and fill it in for each region
                // The file is initially created as empty and the user must load data to it. 
                // The regions can not overlap
                RegionOfInterest r1 = new RegionOfInterest(0, 0, w2[0], h2[0], 1, 1);
                RegionOfInterest r2 = new RegionOfInterest(w2[0]+1, h2[0]+1, w2[1], h2[1], 1, 1);
                RegionOfInterest[] rois2 = { r1,r2 };
                IImageDataSet OneFrameTwoRoi = filemgr.CreateFile(path + "\\OneFrameTwoRoi.spe",
                                                         rois2,
                                                          1,                  // Frames
                                                          ImageDataFormat.MonochromeUnsigned16);

                // Put Data To Region 1
                IImageData data2_roi1 = OneFrameTwoRoi.GetFrame(0, 0);
                data2_roi1.SetData(frame1_roi1);

                // Data To Region 2
                IImageData data2_roi2 = OneFrameTwoRoi.GetFrame(1, 0);
                data2_roi2.SetData(frame1_roi2);

                // Finally close this file
                filemgr.CloseFile(OneFrameTwoRoi);
            }
        }
    }
}
